drivers/vdpa/virtio/virtio_vdpa.c

```diff
@@ -0,0 +1,970 @@
+	irq_set->index = VFIO_PCI_MSIX_IRQ_INDEX;
+	irq_set->start = 0;
+	fd_ptr = (int *)&irq_set->data;
+	fd_ptr[RTE_INTR_VEC_ZERO_OFFSET] =
+		rte_intr_fd_get(priv->pdev->intr_handle);
+
+	for (i = 0; i < nr_vring; i++) {
+		ret = rte_vhost_get_vhost_vring(priv->vid, i, &vring);
+		if (ret){
+			rte_free(irq_set);
+			DRV_LOG(ERR, "Call fd get fail ret:0x%x",ret);
+			return -1;
+		}
+		fd_ptr[RTE_INTR_VEC_RXTX_OFFSET + i] = vring.callfd;
+		DRV_LOG(DEBUG, "Queue %d call fd %d.", i,vring.callfd);
+	}
+
+	ret = ioctl(priv->vfio_dev_fd, VFIO_DEVICE_SET_IRQS, irq_set);
+	rte_free(irq_set);
+	if (ret) {
+		DRV_LOG(ERR, "Error enabling MSI-X interrupts: %s",
+				strerror(errno));
+		return -1;
+	}
+
+	ret = virtio_dev_interrupts_enable(priv->vpdev,nr_vring);
+	if (ret) {
+		DRV_LOG(ERR, "Error enabling virtio dev interrupts: %s",
+				strerror(errno));
+		return -1;
+	}
+	return 0;
+}
+
+static int
+virtio_vdpa_vfio_intr_disable(struct virtio_vdpa_priv *priv)
+{
+	int ret;
+	struct vfio_irq_set irq_set;
+
+	irq_set.argsz = sizeof(struct vfio_irq_set);
+	irq_set.count = 0;
+	irq_set.flags = VFIO_IRQ_SET_DATA_NONE | VFIO_IRQ_SET_ACTION_TRIGGER;
+	irq_set.index = VFIO_PCI_MSIX_IRQ_INDEX;
+	irq_set.start = 0;
+
+	ret = ioctl(priv->vfio_dev_fd, VFIO_DEVICE_SET_IRQS, &irq_set);
+	if (ret) {
+		DRV_LOG(ERR, "Error disabling MSI-X interrupts: %s",
+				strerror(errno));
+		return -1;
+	}
+
+	ret = virtio_dev_interrupts_disable(priv->vpdev, priv->nr_virtqs);
+	if (ret) {
+		DRV_LOG(ERR, "Error disabling virtio dev interrupts: %s",
+				strerror(errno));
+		return -1;
+	}
+
+	return 0;
+}
+
+static void
+virtio_vdpa_virtq_handler(void *cb_arg)
+{
+	struct virtio_vdpa_vring_info *virtq = cb_arg;
+	struct virtio_vdpa_priv *priv = virtq->priv;
+	uint64_t buf;
+	int nbytes;
+
+	if(!(priv->configured))
+		return;
+	if(!virtq->enable)
+		return;
+
+	if (rte_intr_fd_get(virtq->intr_handle) < 0)
+		return;
+
+	do {
+		nbytes = read(rte_intr_fd_get(virtq->intr_handle), &buf, 8);
+		if (nbytes < 0) {
+			if (errno == EINTR ||
+				errno == EWOULDBLOCK ||
+				errno == EAGAIN)
+				continue;
+			DRV_LOG(ERR,  "Failed to read kickfd of virtq %d: %s",
+				virtq->index, strerror(errno));
+		}
+		break;
+	} while (1);
+	virtio_dev_queue_notify(priv->vpdev,virtq->index);
+	if(virtq->notifier_state == VIRTIO_VDPA_NOTIFIER_STATE_DISABLED) {
+		if (rte_vhost_host_notifier_ctrl(priv->vid, virtq->index, true)) {
+			DRV_LOG(ERR,  "Failed to set notify ctrl virtq %d: %s",
+							virtq->index, strerror(errno));
+			virtq->notifier_state = VIRTIO_VDPA_NOTIFIER_STATE_ERR;
+		}
+		else
+			virtq->notifier_state = VIRTIO_VDPA_NOTIFIER_STATE_ENABLED;
+		DRV_LOG(INFO, "Virtq %u notifier state is %s.", virtq->index,
+			virtq->notifier_state ==
+				VIRTIO_VDPA_NOTIFIER_STATE_ENABLED ? "enabled" :
+									"disabled");
+	}
+	DRV_LOG(DEBUG, "Ring virtq %u doorbell.", virtq->index);
+}
+
+static int
+virtio_vdpa_virtq_doorbell_relay_disable(struct virtio_vdpa_priv *priv, int index)
```

Comment:
we always want to work with direct doorbells.

so please remove this code.

Reply:
no, mlx5_vdpa still have this code and it can be used because some time vhost msg come and 
doorbell map still not run.